### PR TITLE
#2384 Built-in Circuit Breaker for Quality of Service (express edition, no Polly required)

### LINKED ISSUE #2384: Circuit breaker pattern
Ocelot uses and recommends [Polly](https://www.pollydocs.org/) for the [Quality of Service](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/qualityofservice.rst) feature.

- Repo: [App-vNext/Polly](https://github.com/App-vNext/Polly)
- Ocelot extension repo: [Ocelot.QualityOfService.Polly](https://github.com/ThreeMammals/Ocelot.QualityOfService.Polly)
- Ocelot extension package: [Ocelot.QualityOfService.Polly](https://www.nuget.org/packages/Ocelot.QualityOfService.Polly/)

## Goal
Use the [Quality of Service](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/qualityofservice.rst) feature without [Polly](https://www.pollydocs.org/) and the [Ocelot.QualityOfService.Polly](https://www.nuget.org/packages/Ocelot.QualityOfService.Polly/) package.

The following acceptance tests must be based on the newly developed Circuit Breaker:
- `test/Ocelot.AcceptanceTests/QualityOfService`
- [DynamicRoutingTests](https://github.com/ThreeMammals/Ocelot/blob/develop/test/Ocelot.AcceptanceTests/ServiceDiscovery/DynamicRoutingTests.cs), especially [ShouldApplyGlobalQosOptions_ForAllDynamicRoutes](https://github.com/ThreeMammals/Ocelot/blob/c3720c550899db4aadd77ae77754bd9978101c34/test/Ocelot.AcceptanceTests/ServiceDiscovery/DynamicRoutingTests.cs#L429) and [ShouldApplyGlobalQosOptions_ForAllDynamicRoutes_WithGroupedOpts](https://github.com/ThreeMammals/Ocelot/blob/c3720c550899db4aadd77ae77754bd9978101c34/test/Ocelot.AcceptanceTests/ServiceDiscovery/DynamicRoutingTests.cs#L463).

## Knowledge Domain
- Wikipedia: [Circuit Breaker design pattern](https://en.wikipedia.org/wiki/Circuit_breaker_design_pattern)
- Martin Fowler: [Circuit Breaker](https://martinfowler.com/bliki/CircuitBreaker.html)
- Microsoft Azure design patterns: [Circuit Breaker pattern](https://learn.microsoft.com/en-us/azure/architecture/patterns/circuit-breaker)
- AWS Cloud design patterns, architectures, and implementations: [Circuit Breaker pattern](https://docs.aws.amazon.com/prescriptive-guidance/latest/cloud-design-patterns/circuit-breaker.html)
- [Microservice Architecture](https://microservices.io/index.html) by Chris Richardson: [Pattern: Circuit Breaker](https://microservices.io/patterns/reliability/circuit-breaker.html)

## Known Top Libraries
- Java: [Netflix Hystrix](https://github.com/Netflix/Hystrix) → [HystrixCircuitBreaker](https://github.com/Netflix/Hystrix/blob/master/hystrix-core/src/main/java/com/netflix/hystrix/HystrixCircuitBreaker.java)
- C# & ASP.NET: [Polly](https://github.com/App-vNext/Polly) → [CircuitBreaker](https://github.com/App-vNext/Polly/tree/main/src/Polly/CircuitBreaker)

## Roadmap
1. Develop a lightweight version of the Circuit Breaker pattern (using the knowledge domain) as a class in the folder `src/Ocelot/Infrastructure/DesignPatterns`. All three main states must be implemented. The initial version should be based on [HystrixCircuitBreaker](https://github.com/Netflix/Hystrix/blob/master/hystrix-core/src/main/java/com/netflix/hystrix/HystrixCircuitBreaker.java), but advanced functionality can be borrowed from [Polly's CircuitBreaker](https://github.com/App-vNext/Polly/tree/main/src/Polly/CircuitBreaker).
2. The implementation class should be based on `System.Net.Http.DelegatingHandler` since Ocelot's [MessageInvokerPool](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Requester/MessageInvokerPool.cs) creates a `SocketsHttpHandler` object for each route. This handler is then wrapped by `TimeoutDelegatingHandler` as the top handler to create an `HttpMessageInvoker` to send requests. Therefore, we cannot use `HttpClient` directly.
3. Reuse the current options for the [Quality of Service](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/qualityofservice.rst) feature model, i.e. the [QoSOptions](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Configuration/QoSOptions.cs) class. The core QoS feature functionality and core services are located in the [QualityOfService](https://github.com/ThreeMammals/Ocelot/tree/develop/src/Ocelot/QualityOfService) folder.
4. Write unit tests (xUnit, using Assert-class based methods) in the folder: `test/Ocelot.UnitTests/Infrastructure/DesignPatterns`.
5. Write acceptance tests for basic HTTP methods, then tests for the main Circuit Breaker cases.
6. Rewrite the acceptance tests for classes utilizing the developed circuit breaker: [ShouldApplyGlobalQosOptions_ForAllDynamicRoutes](https://github.com/ThreeMammals/Ocelot/blob/c3720c550899db4aadd77ae77754bd9978101c34/test/Ocelot.AcceptanceTests/ServiceDiscovery/DynamicRoutingTests.cs#L429) and [ShouldApplyGlobalQosOptions_ForAllDynamicRoutes_WithGroupedOpts](https://github.com/ThreeMammals/Ocelot/blob/c3720c550899db4aadd77ae77754bd9978101c34/test/Ocelot.AcceptanceTests/ServiceDiscovery/DynamicRoutingTests.cs#L463). The `Skip` flag must be removed from the `Fact` attributes to re-enable the tests, which must pass.
7. Optionally, update the documentation to mention the new Circuit Breaker strategy, but without referencing Polly. This will be built-in Ocelot functionality.

## Predecessors
- #2380 
- #561

### PR DESCRIPTION
**Pull request by ocelotgateway**
## Closes #2384 
- #2384 

## Predecessors
- #2380 
- #561 

## Proposed Changes
Ocelot's QoS feature previously had no built-in circuit breaker implementation — it depended entirely on the external `Ocelot.QualityOfService.Polly` package. This adds a lightweight, built-in circuit breaker so QoS works out of the box.

## Core implementation

- **`src/Ocelot/Infrastructure/DesignPatterns/CircuitBreaker.cs`** — Thread-safe state machine (`Closed → Open → HalfOpen → Closed`) inspired by HystrixCircuitBreaker. `MinimumThroughput ≤ 0` disables circuit-opening while still allowing timeout enforcement.

- **`src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs`** — `DelegatingHandler` that integrates with `MessageInvokerPool`'s handler chain. Counts 5xx responses and non-cancellation exceptions as failures. Enforces per-request timeout via `QoSOptions.Timeout` using a linked `CancellationTokenSource`. Key subtlety: `OperationCanceledException` is re-thrown without recording a failure in the inner `SendAndTrackAsync`, preventing double-counting when the QoS timeout fires.

```csharp
// OcelotBuilder now registers a default — overridable by external packages via AddSingleton after AddOcelot()
Services.TryAddSingleton<QosDelegatingHandlerDelegate>(provider =>
    (route, contextAccessor, loggerFactory) => new CircuitBreakerDelegatingHandler(route, loggerFactory));
```

## DI registration

- **`OcelotBuilder.cs`** — Registers `CircuitBreakerDelegatingHandler` as the default `QosDelegatingHandlerDelegate` via `TryAddSingleton`. Polly (or any other QoS package) can still override by registering after `AddOcelot()`.

## Tests

- **`CircuitBreakerTests.cs`** (new, 12 tests) — Unit tests covering all state transitions and edge cases including zero/negative `MinimumThroughput`.

- **`DynamicRoutingTests.cs`** — Removed `Skip` from `ShouldApplyGlobalQosOptions_ForAllDynamicRoutes` and `ShouldApplyGlobalQosOptions_ForAllDynamicRoutes_WithGroupedOpts`. Removed the `FakeQualityOfServiceDelegatingHandler` override from `WithDiscoveryAndQualityOfService` — both tests now run against the real built-in handler and pass.

---
## Proposed Changes during Dev Round 2

**`src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs`**
- Added `LowTimeout = 10`, `DefaultTimeout = 30_000`, `HighTimeout = 86_400_000` constants aligned with Polly's `TimeoutStrategy`
- Renamed `public static readonly ServerErrorCodes` &rarr; `DefaultServerErrorCodes`
- Added `protected virtual HashSet<HttpStatusCode> ServerErrorCodes` property (defaults to `DefaultServerErrorCodes`) — subclasses can override to customise failure codes
- Added `GetTimeout()` helper: null/≤0 &rarr; disabled, out-of-Polly-range &rarr; `DefaultTimeout`, valid &rarr; value as-is
- Stored result in `_timeoutMs` field; `SendAsync` and `SendWithTimeoutAsync` use `_timeoutMs` instead of reading `_options.Timeout` directly

**`src/Ocelot/DependencyInjection/IOcelotBuilder.cs` + `OcelotBuilder.cs`**
- Added `AddQualityOfService<THandler>()` generic overload where `THandler : CircuitBreakerDelegatingHandler`; registers a `QosDelegatingHandlerDelegate` factory using `ActivatorUtilities.CreateInstance<THandler>`, enabling custom handler subclasses to be injected via DI

**`test/Ocelot.UnitTests/QualityOfService/CircuitBreakerDelegatingHandlerTests.cs`**
- Updated `ServerErrorCodes` references &rarr; `DefaultServerErrorCodes`
- Added constant tests for `LowTimeout`, `DefaultTimeout`, `HighTimeout`
- Added `GetTimeout` boundary tests (disabled values, boundary values using default)
- Added `CustomErrorCodesHandler` inner subclass and two tests proving the virtual `ServerErrorCodes` override works correctly

**`docs/features/qualityofservice.rst`**
- **Table row MinimumThroughput=0:** Swapped columns — Built-in now "Uses default (100)", Polly now "Disables circuit breaking"
- **Timeout constraints row:** Updated valid range to `> 10 ms, < 86 400 000 ms`; default to `30 000 ms`; notes updated
- **Timeout section (line 166):** Added `.. note::` explaining that timeout-only config still activates the circuit breaker with defaults (`MinimumThroughput=100`, `BreakDuration=5000ms`) and how to disable the timeout
- **Server Error Codes section (line 205):** Added "Overriding server error codes" sub-section with a C# subclassing example and the `AddQualityOfService<MyCircuitBreakerHandler>()` registration snippet

<details><summary><h2>Development History Summary</h2></summary>

This PR was built across **9 phases**, alternating between Copilot agent sessions and manual review/polish commits by @ocelotgateway.

### Phase 1 — Initial Implementation (Apr 20, 2026) · *Copilot agent session 1*

- **Initial plan** commit established the work breakdown.
- Introduced the core `CircuitBreaker` class with `Closed`/`Open`/`HalfOpen` state machine in **count mode** (`MinimumThroughput` consecutive failures).
- Added `CircuitBreakerDelegatingHandler` — the HTTP pipeline handler recording successes/failures, enforcing timeout via `CancellationToken`, and returning `503` when open.
- Wired up the built-in handler into `QualityOfServiceFactory` and registered `AddQualityOfService()` on `OcelotBuilder`.

---

### Phase 2 — Manual Review & Refactoring (Apr 21–22, 2026) · *@ocelotgateway / Ocelot Robo*

- `IQoSFactory` → `IQualityOfServiceFactory`, `QosFactory` → `QualityOfServiceFactory`.
- Reorganised DI registration into `Features.cs` (`AddOcelotQualityOfService()`).
- Added `WaitForState` helper to `CircuitBreakerTests`.
- Added acceptance tests; placeholder `FailureRatio` tests marked skipped.

---

### Phase 3 — Bug Fixes & Full Unit Test Coverage (Apr 22, 2026) · *Copilot agent [session 6504f59b](https://github.com/ocelotgateway/Ocelot/sessions/6504f59b-647b-4adc-9263-a480c32d6120)*

- Fixed `GetBreakDuration`: `>=` → `>` so `BreakDuration=500` (exact boundary) uses the 5 000 ms default.
- Fixed `QualityOfServiceFactory.Get` return path for the built-in handler.
- Added comprehensive unit tests for all `CircuitBreakerDelegatingHandler` behaviours.

---

### Phase 4 — Missing Coverage (Apr 23, 2026) · *Copilot agent sessions [f1cfc680](https://github.com/ocelotgateway/Ocelot/sessions/f1cfc680-b95f-4c1f-8a4a-b47e69f922d5) · [4879cba4](https://github.com/ocelotgateway/Ocelot/sessions/4879cba4-a01d-46e9-986e-42cf8294fcaa)*

- Covered `QualityOfServiceFactory` null-`HttpContext` branches (lines 36 & 48).
- Covered `OcelotBuilder.AddQualityOfService()` lines 341–342.
- @ocelotgateway applied follow-up code-review suggestions.

---

### Phase 5 — FailureRatio Mode (Apr 23, 2026) · *Copilot agent sessions [76ebb7d9](https://github.com/ocelotgateway/Ocelot/sessions/76ebb7d9-9049-4419-b67a-fc11f9a4a154) · [3095df42](https://github.com/ocelotgateway/Ocelot/sessions/3095df42-f787-4952-bac6-db743f315162)*

- Extended `CircuitBreaker` with **ratio mode**: rolling-window queue, `PurgeOldEntries()`, opens at `windowCount >= MinimumThroughput && failures/total >= FailureRatio`.
- `CircuitBreakerDelegatingHandler` auto-selects count vs ratio mode.
- Acceptance tests for `FailureRatio × MinimumThroughput` interaction.

---

### Phase 6 — Final Polish (Apr 25, 2026) · *@ocelotgateway / Ocelot Robo*

- Removed `QosDelegatingHandlerDelegate` indirection; factory creates handler directly.
- Improved `FileQoSOptionsFluentValidator` error message for both implementations.
- Refactored `QosSteps` constructors; improved thread safety; updated log/XML-doc comments.

---

### Phase 7 — HalfOpen Thundering-Herd Fix (Apr 25, 2026) · *Copilot agent [session b5e4867a](https://github.com/ocelotgateway/Ocelot/sessions/b5e4867a-1ace-4f29-9ce9-2226dfb45402)*

- Added `_halfOpenProbeInFlight` boolean gate: exactly **one** probe allowed through in `HalfOpen`; all other concurrent calls return `503` immediately.
- Unit tests asserting single-probe semantics and concurrent-call rejection.

---

### Phase 8 — CircuitBreaker Correctness Fixes (Apr 25, 2026) · *Copilot agent [session 99a3a78a](https://github.com/ocelotgateway/Ocelot/sessions/99a3a78a-b9f2-456a-a512-56908dc6490d)*

1. **Removed `volatile` from `_state`** — all reads/writes are inside `lock (_lock)`; `volatile` was incorrect.
2. **Added `MinimumThroughput <= 0` guard in `RecordFailureRatioMode`** — prevents ratio mode opening the circuit on the first failure when throughput is zero.
3. **Eliminated wasted enqueue in `RecordSuccess` ratio/HalfOpen path** — restructured to branch before enqueuing.

---

### Phase 9 — Documentation (Apr 25, 2026) · *Copilot agent [session 833a6cc4](https://github.com/ocelotgateway/Ocelot/sessions/833a6cc4-7b89-4378-86c4-fe000a47a02c)*

Updated `docs/features/qualityofservice.rst`:
- New **Implementations Overview** comparison table (built-in vs Polly).
- New **Built-in QoS** section: activation, count mode, ratio mode, state-machine table, timeout, server error codes, value constraints.
- Polly sections annotated `(Polly)` with cross-references to built-in equivalents.
- All existing RST cross-reference labels preserved.

</details>

### CODE REVIEW

File: src/Ocelot/QualityOfService/QualityOfServiceFactory.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        if (handler is not QosDelegatingHandler)
+            return handler; // this is external lib like Polly
+
+        // Built-in Quality of Service feature was enabled by AddOcelot().AddQualityOfService()
+        if (opts.MinimumThroughput.HasValue || opts.BreakDuration.HasValue)
+            return new CircuitBreakerDelegatingHandler(route, _loggerFactory);
+
+        // TODO Timeout strategy, but it requires refactoring of MessageInvokerPool and TimeoutDelegatingHandler
+        //if (opts.Timeout.HasValue)
+        //    return new TimeoutDelegatingHandler(TimeSpan.FromMilliseconds(opts.Timeout.Value));
```
[the remark this answers is not in the document]
Comment (ocelot-ot) [reply src/Ocelot/QualityOfService/QualityOfServiceFactory.cs]: It is a subject of design. Will be fixed later as stated in the TODO

File: src/Ocelot/Infrastructure/DesignPatterns/CircuitBreaker.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                }
+            }
+
+            return _state;
+        }
+    }
+
+    /// <summary>Returns <see langword="true"/> if a request can proceed (circuit is <see cref="CircuitState.Closed"/> or <see cref="CircuitState.HalfOpen"/>).</summary>
+    public bool CanExecute() => State != CircuitState.Open;
```
[the remark this answers is not in the document]
Comment (ocelot-ot) [reply src/Ocelot/Infrastructure/DesignPatterns/CircuitBreaker.cs]: I will rethink this. Need to check "the HalfOpen documentation"...

[the remark this answers is not in the document]
Comment (ocelot-ot) [reply src/Ocelot/Infrastructure/DesignPatterns/CircuitBreaker.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/9a19241e9dd71b07b63154fab277273586fb1ea9 ✔️

File: docs/features/qualityofservice.rst [hunk 3]
```diff
@@ -17,21 +17,237 @@ Quality of Service
...
+      - ✔
+    * - Extensibility API
+      - ✘
+      - ✔ (custom providers, handlers, error maps)
+    * - Invalid-value handling
+      - Silent default substitution
+      - Logged warning + default substitution
+    * - ``MinimumThroughput = 0``
+      - Disables circuit breaking
+      - Uses default (100)
```
Comment (raman-m) [inline docs/features/qualityofservice.rst]: Look at this UseQos guard:
https://github.com/ocelotgateway/Ocelot/blob/copilot/implement-circuit-breaker-pattern/src/Ocelot/Configuration/QoSOptions.cs#L64-L66
For Polly implementation it also disables Circuit Breaker due to this line of code:
https://github.com/ThreeMammals/Ocelot.QualityOfService.Polly/blob/main/src/PollyQoSResiliencePipelineProvider.cs#L76
Due to validation check default value is substituted only if validation has failed:
https://github.com/ThreeMammals/Ocelot.QualityOfService.Polly/blob/main/src/PollyQoSResiliencePipelineProvider.cs#L84
Finally, this thesis is false. 👇 

```suggestion
      - Disables circuit breaking
```

Additionally, we should say about Timeout strategy disabling aka "how to".

File: docs/features/qualityofservice.rst [hunk 4]
```diff
@@ -17,21 +17,237 @@ Quality of Service
...
+      - 0.5
+      - Ratio mode only.
+    * - ``SamplingDuration``
+      - > 500 ms
+      - 10 000 ms
+      - Ratio mode only.
+    * - ``Timeout``
+      - > 0 ms
+      - (disabled)
+      - Any value ≤ 0 disables the timeout.
```
Comment (raman-m) [inline docs/features/qualityofservice.rst]: This is wrong! This is design issue.
Timeout value range should be adjusted to Polly's constants: https://github.com/ThreeMammals/Ocelot.QualityOfService.Polly/blob/main/src/TimeoutStrategy.cs
Thus, a default value must be introduced in the `CircuitBreakerDelegatingHandler` constructor.

File: src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    public const double DefaultFailureRatio = 0.5;
+
+    public const int LowSamplingDuration = 500;
+    public const int DefaultSamplingDuration = 10_000;
+
+    /// <summary>
+    /// The set of HTTP status codes that are considered server errors and will be counted as circuit-breaker failures.
+    /// </summary>
+    /// <remarks>Used by Ocelot's built-in QoS handler to treat downstream 5xx responses as circuit-breaker failures.</remarks>
+    public static readonly IReadOnlySet<HttpStatusCode> ServerErrorCodes = new HashSet<HttpStatusCode>
```
Comment (raman-m) [inline src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs]: It should be default set of codes, and it should be assigned by default.
Additional protected virtual property `ServerErrorCodes` should be introduced to allow a developer to inherit and override those codes, see Polly's implementation: 
https://github.com/ThreeMammals/Ocelot.QualityOfService.Polly/blob/d77e771592e478b178939babfbd45164dc4e6547/src/PollyQoSResiliencePipelineProvider.cs#L39

File: docs/features/qualityofservice.rst [hunk 6]
```diff
@@ -17,21 +17,237 @@ Quality of Service
...
+Timeout
+^^^^^^^
+
+An optional per-request timeout can be configured independently of or alongside the circuit breaker:
+
+.. code-block:: json
+
+  "QoSOptions": {
+    "Timeout": 5000
+  }
```
Comment (raman-m) [inline docs/features/qualityofservice.rst:173]: With this option, we set up a timeout which is part of the circuit breaker where the timeout is implemented.
We need to explain to the user that the default values of the built-in implementation are substituted here.
By design, the timeout feature is not independent of the circuit breaker; thus, it is part of it.
Therefore, an explanation is required that the default values are substituted with such a configuration.

File: docs/features/qualityofservice.rst [hunk 7]
```diff
@@ -17,21 +17,237 @@ Quality of Service
...
+      - HTTP Version Not Supported
+    * - 506
+      - Variant Also Negotiates
+    * - 507
+      - Insufficient Storage
+    * - 508
+      - Loop Detected
+
+Any other status code (including ``4xx`` client errors) is recorded as a success and does not contribute to the failure count.
+Unhandled exceptions (excluding ``OperationCanceledException``) are also counted as failures.
```
Comment (raman-m) [inline docs/features/qualityofservice.rst:221]: We have to explain how to override the codes virtually.

Before this doc update, we must introduce a new DI helper in `OcelotBuilder` to inject a descendant class inheriting from `CircuitBreakerDelegatingHandler`. Thus, the descendant handler will override the codes.

**Review by raman-m — CHANGES_REQUESTED** [review]
The design is good, but I believe a few final tweaks should be applied. See my suggestions below.

**Comment by ocelotgateway** [comment]
Hey @raman-m **Done!**
Details are here:
- [x] Read review comments from raman-m (review #4189528119)
- [x] `CircuitBreakerDelegatingHandler.cs`: add `LowTimeout`/`DefaultTimeout`/`HighTimeout` constants; rename `ServerErrorCodes` static field to `DefaultServerErrorCodes`; add `protected virtual HashSet<HttpStatusCode> ServerErrorCodes` property; add `GetTimeout` helper storing validated value in `_timeoutMs`; use `_timeoutMs` in `SendAsync` and `SendWithTimeoutAsync`
- [x] `IOcelotBuilder.cs` + `OcelotBuilder.cs`: add generic `AddQualityOfService<THandler>()` overload (where `THandler : CircuitBreakerDelegatingHandler`) using `ActivatorUtilities.CreateInstance`
- [x] `CircuitBreakerDelegatingHandlerTests.cs`: renamed `ServerErrorCodes` → `DefaultServerErrorCodes` in test assertions; added tests for `LowTimeout`, `DefaultTimeout`, `HighTimeout`; added tests for `GetTimeout` boundary behavior; added tests for virtual `ServerErrorCodes` override (`CustomErrorCodesHandler`)
- [x] `docs/features/qualityofservice.rst`: swapped MinimumThroughput=0 table row (Built-in now "Uses default (100)", Polly now "Disables circuit breaking"); fixed Timeout constraints row (range > 10 ms / < 86 400 000 ms, default 30 000 ms); added note about default values substituted in timeout-only config; added "Overriding server error codes" sub-section with code example for subclassing + `AddQualityOfService<THandler>()` registration

File: docs/features/qualityofservice.rst (hunk 3 above)
Comment (ocelot-ot) [reply docs/features/qualityofservice.rst]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/1877af28f42dc4267c27f4c6ee00564b61cbc4f3

File: docs/features/qualityofservice.rst (hunk 4 above)
Comment (ocelot-ot) [reply docs/features/qualityofservice.rst]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/1877af28f42dc4267c27f4c6ee00564b61cbc4f3

File: src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs (hunk 5 above)
Comment (ocelot-ot) [reply src/Ocelot/QualityOfService/CircuitBreakerDelegatingHandler.cs]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/1877af28f42dc4267c27f4c6ee00564b61cbc4f3

File: docs/features/qualityofservice.rst (hunk 6 above)
Comment (ocelot-ot) [reply docs/features/qualityofservice.rst:173]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/1877af28f42dc4267c27f4c6ee00564b61cbc4f3

File: docs/features/qualityofservice.rst (hunk 7 above)
Comment (ocelot-ot) [reply docs/features/qualityofservice.rst:221]: Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/2385/commits/1877af28f42dc4267c27f4c6ee00564b61cbc4f3

**Review by raman-m — APPROVED** [review]
## Ready for delivery ✅ 
- Code review ✔️ ✔️ ✔️ 
- Unit testing ✔️ ✔️ ✔️ 
- Acceptance testing ✔️ ✔️ 
- Updated docs ✔️ ✔️
